### DISCUSSION TITLE
Beneficiary Deduplication

### DISCUSSION THREAD

**Discussion by manjik-rumsan** [rahataid/rahat-platform#115] [Ideas]
We need to implement a deduplication mechanism in our beneficiary management tool. Currently, extra fields for beneficiaries are defined dynamically by users and stored as JSON in the database. This setup makes it challenging to efficiently identify and remove duplicate entries.

Based on the internal discussion with @santosh-rumsan, we are proposing the following solution:

- **Dynamic Table Creation:** Create a relational database table dynamically based on the user-defined field definitions.
- **pg_trgm Extension:** Leverage the `pg_trgm` extension in PostgreSQL to perform similarity-based text searches and calculate the similarity score. This will allow us to identify potential duplicates even when text fields (e.g., names, addresses, phones) have slight variations.
- **Review Process:** Implement a review process for identified duplicates, allowing users to merge or discard duplicate records.

# Here is the example suggested by CHATGPT
### Example: Deduplication Based on Multiple Fields

Suppose you want to deduplicate records based on name, email, and address.

#### 1. Enable pg_trgm

```sql
CREATE EXTENSION IF NOT EXISTS pg_trgm;
```

#### 2. Create the PostgreSQL Query

The following SQL query calculates a combined similarity score by averaging the individual similarity scores of `name`, `email`, and `address`. You can then filter records where the combined score exceeds a certain threshold.

```sql
SELECT
  id,
  name,
  email,
  address,
  (similarity(name, $1) + similarity(email, $2) + similarity(address, $3)) / 3 AS combined_similarity
FROM
  users
WHERE
  (similarity(name, $1) + similarity(email, $2) + similarity(address, $3)) / 3 > $4
ORDER BY
  combined_similarity DESC;
```

#### 3. TypeScript Implementation

Here’s how you can integrate this into your TypeScript code:

```typescript
import { Pool } from 'pg';

// Set up your database connection
const pool = new Pool({
  user: 'your_username',
  host: 'localhost',
  database: 'your_database',
  password: 'your_password',
  port: 5432,
});

// Function to get similar records based on multiple fields
async function findPotentialDuplicates(record: { name: string, email: string, address: string }, threshold: number) {
  const query = `
    SELECT
      id,
      name,
      email,
      address,
      (similarity(name, $1) + similarity(email, $2) + similarity(address, $3)) / 3 AS combined_similarity
    FROM
      users
    WHERE
      (similarity(name, $1) + similarity(email, $2) + similarity(address, $3)) / 3 > $4
    ORDER BY
      combined_similarity DESC;
  `;

  const client = await pool.connect();
  try {
    const result = await client.query(query, [record.name, record.email, record.address, threshold]);
    return result.rows;
  } finally {
    client.release();
  }
}

// Deduplication logic
async function deduplicateRecords(records: { name: string, email: string, address: string }[]) {
  for (const record of records) {
    const potentialDuplicates = await findPotentialDuplicates(record, 0.3); // Use 0.3 as the threshold
    console.log(`Record: ${record.name}, Potential Duplicates:`, potentialDuplicates);
  }
}

// Sample records to check for duplicates
const records = [
  { name: 'John Doe', email: 'john.doe@example.com', address: '123 Main St' },
  { name: 'Jon Doe', email: 'jon.d@example.com', address: '123 Main St' },
  { name: 'Jane Doe', email: 'jane.doe@example.com', address: '456 Elm St' },
];

deduplicateRecords(records)
  .then(() => console.log('Deduplication completed'))
  .catch((err) => console.error('Error in deduplication:', err));
```

### Explanation

1. **Combined Similarity Score:** The query calculates a `combined_similarity` by averaging the similarity scores of `name`, `email`, and `address`. You could also use a weighted average if some fields are more important than others.
2. **Threshold:** The threshold (0.3 in the example) applies to the combined similarity score. Adjust this to tune the sensitivity of the deduplication process.
3. **Performance Considerations:** As with single-field deduplication, you can create a GIN index on each of the fields involved to optimize the search.

**Comment by manjik-rumsan** [comment]
@binod-rumsan what do you think of this approach?

**Comment by manjik-rumsan** [comment]
@avash-rumsan @raghav-rumsan @anupama-rumsan 
Love to hear some ideas/suggestions

**Comment by santosh-rumsan** [comment]
Love the nice description of this issue 😀
